fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/InliningStorageAdapter.java

```diff
@@ -0,0 +1,383 @@
+     *
+     * @param readTransaction the transaction to use for reading data
+     * @param layer the layer of the graph to scan
+     * @param lastPrimaryKey the primary key of the last node read in a previous scan, used for pagination.
+     * If {@code null}, the scan starts from the beginning of the layer.
+     * @param maxNumRead the maximum number of raw key-value records to read from the database
+     * @return an {@code Iterable} of {@link AbstractNode} objects reconstructed from the scanned layer. Each node
+     *         contains its neighbors within that layer.
+     */
+    @Nonnull
+    @Override
+    public Iterable<AbstractNode<NodeReferenceWithVector>> scanLayer(@Nonnull final ReadTransaction readTransaction,
+                                                                     int layer,
+                                                                     @Nullable final Tuple lastPrimaryKey,
+                                                                     int maxNumRead) {
+        final OnReadListener onReadListener = getOnReadListener();
+        final byte[] layerPrefix = getDataSubspace().pack(Tuple.from(layer));
+        final Range range =
+                lastPrimaryKey == null
+                ? Range.startsWith(layerPrefix)
+                : new Range(ByteArrayUtil.strinc(getDataSubspace().pack(Tuple.from(layer, lastPrimaryKey))),
+                        ByteArrayUtil.strinc(layerPrefix));
+        final AsyncIterable<KeyValue> itemsIterable =
+                readTransaction.getRange(range,
+                        maxNumRead, false, StreamingMode.ITERATOR);
+        Tuple nodePrimaryKey = null;
+        ImmutableList.Builder<AbstractNode<NodeReferenceWithVector>> nodeBuilder = ImmutableList.builder();
+        ImmutableList.Builder<NodeReferenceWithVector> neighborsBuilder = null;
+        for (final KeyValue item: itemsIterable) {
+            final byte[] key = item.getKey();
+            final byte[] value = item.getValue();
+            onReadListener.onKeyValueRead(layer, key, value);
+
+            final Tuple neighborKeyTuple = getDataSubspace().unpack(key);
+            final Tuple neighborValueTuple = Tuple.fromBytes(value);
+            final NodeReferenceWithVector neighbor =
+                    neighborFromTuples(AffineOperator.identity(), neighborKeyTuple, neighborValueTuple);
+            final Tuple nodePrimaryKeyFromNeighbor = neighborKeyTuple.getNestedTuple(1);
+            if (nodePrimaryKey == null || !nodePrimaryKey.equals(nodePrimaryKeyFromNeighbor)) {
+                if (nodePrimaryKey != null) {
+                    nodeBuilder.add(getNodeFactory().create(nodePrimaryKey, null, neighborsBuilder.build()));
+                }
+                nodePrimaryKey = nodePrimaryKeyFromNeighbor;
+                neighborsBuilder = ImmutableList.builder();
+            }
+            neighborsBuilder.add(neighbor);
+        }
+
+        // there may be a rest; throw it away
```

Comment:
Does this not have the problem that the last node in the layer will never get added to the list? The idea being that because it always ignores the final contents of `neighborsBuilder` (because it doesn't know if there are more neighbors or not), it will never return the final node in the layer